man/im_open.3
.TH IM_OPEN 3 "30 October 1992"
.SH NAME
im_open, im_open_local, im_open_local_array \- open VIPS 
image descriptor(s)
.SH SYNOPSIS
#include <vips/vips.h>

IMAGE *im_open( const char *filename, const char *mode )

IMAGE *im_open_local( IMAGE *im, const char *filename, const char *mode )

int im_open_local_array( IMAGE *im, 
  IMAGE **out, int n, const char *filename, const char *mode )

.SH DESCRIPTION
.B im_open(3)
examines the mode string, and creates an appropriate VIPS IMAGE descriptor.

.B "r" 
opens the named file for reading. If the file is not in the native VIPS format
for your machine,
.B im_open(3)
automatically converts the file for you in memory. For some large files (eg.
TIFF) this may not be what you want: you should call the appropriate converter
yourself, and arrange for the conversion to take place on disc. See
.B im_tiff2vips(3),
.B im_jpeg2vips(3),
.B im_png2vips(3),
.B im_magick2vips(3),
and
.B im_ppm2vips(3).

.B im_open(3)
can read files in most formats.

.B "w"
opens the named file for writing. It looks at the file name suffix to
determine the type to write -- for example:

  im_open( "fred.tif", "w" )

will write in TIFF format.

You can pass parameters to the conversion functions encoded in the filename
string. For example:

  im_open( "fred.tif:deflate", "w" )

will write a deflate (ZIP) compressed TIFF file. See the man pages for
.B im_vips2tiff(3),
.B im_vips2jpeg(3),
.B im_vips2png(3)
and
.B im_vips2ppm(3)
for details on all of the options available.

.B "t"
creates a temporary memory buffer image.

.B "p"
creates a "glue" descriptor you can use to join two image processing
operations together.

.B "rw"
opens the named file for reading and writing. This will only work for VIPS
files in a format native to your machine. It is only for paintbox-type
applications.

.B im_open_local(3) 
is a convenience function which opens an image descriptor as
im_open(3), but makes it local to im, that is, when im is closed, the
descriptor created by im_open_local(3) will be closed too.

.B im_open_local(3) 
is handy for saving you from adding many 
.B im_close(3) 
calls to
escape points. Example: find the total of an array of images.

  #include <vips/vips.h>

  int
  total( IMAGE **in, int nin, IMAGE *out )
  {
    int i;
    IMAGE *t1, *t2;

    if( nin <= 0 ) {
      im_errormsg( "total: nin should be > 0" );
      return( -1 );
    }
    else if( nin == 1 )
      return( im_copy( *in, out ) );
    else
      for( t1 = *in, i = 1; i < nin; i++ ) {
        if( i + 1 == nin )
	  t2 = out;
	else if( !(t2 = im_open_local( out, "t2", "p" )) )
          return( -1 );

        if( im_add( t1, in[i], t2 ) )
          return( -1 );
        t1 = t2;
      }

    return( 0 );
  }

This function will create many intermediate images, but does not need to close
them. Any which are created will be closed automatically when out is closed by
our caller.

.B im_open_local(3) 
returns NULL on error, or if its first parameter is NULL.

.B im_open_local_array(3)
will open an array of images, failing if any of the opens fail. It's handy if
you need a number of images for intermediates. Example:

  IMAGE *t[6];

  if( im_open_local_array( out, t, 6, "mytemps", "p" ) )
    return( -1 );

opens 6 temp images (t[0] to t[5]).

.SH RETURN VALUE
The function returns the image descriptor on success and NULL on error.
.SH SEE ALSO
im_close(3), im_vips2tiff(3), im_vips2jpeg(3), im_vips2ppm(3),
im_tiff2vips(3), im_jpeg2vips(3), im_ppm2vips(3).
.SH COPYRIGHT
K. Martinez, 1992.
.SH AUTHOR
K. Martinez.


NEWS
VIPS changed from 7.10 to 7.12 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- IMAGE->Hist is deprecated, use im_history_get() instead
- im_append_Hist() removed
- IMAGE->Bbits is deprecated (now ignored)
- im_region_local() replaced by im_region_buffer()

VIPS enhancements
=================

- new backwards and forwards compatible VIPS file format: it has a new 
  metadata system that efficiently shares and propogates ICC profiles,
  EXIF data, etc. & whatever
- speed ups: new pixel buffer cache saves and reuses computations, uses liboil 
  where possible for a small speedup
- SMP scaling: simplified lock system improves SMP scaling, double-buffered 
  image writes overlap computation 
- lower mem use: new mmap window cache shares file mappings, history buffers 
  share and reuse image history metadata
- built-in support for new image formats: OpenEXR, Analyze7, CSV
- RGB16 and GREY16 image hints
- uses GOption for much better command-line argument handling
- better C++ binding: namespaces, VError derives from std::exception, image
  and number vector constants
- python binding 
- gcc sentinel attributes added
- added GValue as an arg type
- added im_benchmark() and friends for testing
- new functions: 
	im_analyze2vips(), im_benchmark*(), im_buildlut(),
	im_concurrency_get/set(), im_contrast_surface(),
	im_contrast_surface_raw(), im_copy_from(), im_copy_morph(),
	im_cp_desc_array(), im_cp_descv(), im_extract_areabands(),
	im_flood_blob_copy(), im_get_option_group(), im_grid(),
	im_header_exists(), im_header_map(), im_history_get(),
	im_invalidate(), im_isscalar(), im_lineset(), im_linreg(),
	im_meta*(), im_msb(), im_norm_dmask(), im_project(),
	im_raw2vips(), IM_REGION_ADDR_TOPLEFT(), im_tile_cache(),
	im_video_test()


VIPS changed from 7.8 to 7.10 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- no longer generates vips-config script, instead it makes .pc files for
  pkg-config ... replace any "vips-config" with "pkg-config vips-7.10"
- origin hint semantics changed ... it now records the position in the output
  image of the origin of the input image
- all float to int casts on pixel values are now floor() rather than rint(),
  for consistency with C casting semantics

VIPS enhancements
=================

- nip reworked for gtk+-2.4, now called nip2
- now based on glib, so VIPS no longer has it's own thread, plugin and data
  structure libraries
- i18n support, although not quite complete yet
- new error message API to make i18n possible (compatibility macros mean the
  old API still works)
- new 'start up VIPS' call, im_init_world(), sets up i18n and threads
  (optional to call it, but i18n won't fully work unless you do)
- C programs now just need "#include <vips/vips.h>", no other includes
  required
- wraps the Pango font rendering library, see im_text()
- new function flag: NOCACHE ... means the result of the call should not be
  cached, useful for video grab functions and paintbox functions
- im_embed() now supports a wider range of embedding styles, including a fast
  edge stretch
- all area operations use the new embed style to expand their input rather 
  than their output, so there are no more black borders on filtered images
- other new functions: im_render(), im_cache(), im_extract_bands(), 
  im_copy_swap(), im_rint(), im_make_xy(), im_init_world(), im_error(),
  im_warn(), im_diag(), im_rank_image(), im_replicate()
- many fixes and improvements to old funcs
- configure and portability improvements

Bug fixes
=========

- all serious bug fixes got back-ported to 7.8, I think


VIPS changes from 7.6 to 7.8 (again, not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- output format options specified in filenames: new prototypes for load and
  save tiff and jpeg
- C++ API name changes ... error => VError
- include path change
    - C includes <vips/vips.h>
    - C++ includes <vips/vipscpp.h>
- im_extract*() numbers bands from zero, not from 1 ... also other funcs which
  refer to bands (eg. im_lrmosaic() etc.)
- mosaic funcs have extra max blend width parameter

VIPS enhancements
=================

- rewritten and much fancier ip (see separate docs), now called nip ... old ip
  (slightly fixed) still there
- mmap windows remove 2GB image size limit
- images have an origin hint field
- icc transforms supported via wrapping of Little CMS
- much, much faster Fourier transforms via wrapping of libfwfft
- ppm/pgm/pbm read/write
- C++ API enhancements: new constructors, more operator equivalences
- new colourspace: Yxy
- support for D50 colour temperature images
- new functions: im_image(), im_system(), im_version*(), im_blend(),
  im_copy_set(), im_flood_blob(), im_icc_*(), im_open_local_array(),
  im_header_*(), im_sign(), im_ceil(), im_floor(), im_remainderconst*(),
  im_print(), im_guess_prefix(), im_remosaic(), im_invertlut(), Yxy funcs,
  color temperature funcs, im_clip2fmt(), im_blend(), im_lab_morph(),
  im_histnorm(), im_histcum(), im_video_v4l(), im_LabS2Lab(), im_Lab2LabS()
- new type: IMAGEVEC
- header is much faster
- ip/nip split to separate packages
- better vips2dj
- better cygwin support, mingw support too
- Mac OS X and Windows build support
- new set of sensible #define names, old names can be turned off with
  -DIM_NO_VIPS7_COMPAT
- many configure improvements

Bug fixes
=========

- speed ups
- fixes to implicit image format conversions 
- im_zoom() spots integer overflow
- im_LabS2LabQ() rounding on a/b slightly broken for a/b == 0
- fixes to refcounting in C++ API
- mask casts in C++ were broken




VIPS Changes for 7.4 to 7.6

Non-backwards compatible changes
================================

- im_histplot() has new (and better) rules
- im_andconst(), im_orconst() and im_eorconst() now take a
  double, not an unsigned char, as their constant argument type

libsrc/iofuncs/im_open.c
/* @(#) im_open: vips front end to file open functions
ARGS: filename and mode which is one of:
"r" read by mmap (im_mmapin)
"rw" read and write by mmap (im_mmapinrw)
"w" write for write (im_writeline)
"t" for temporary memory image
"p" for partial memory image
RETURNS: IMAGE pointer or 0 on error
Copyright: Kirk Martinez 29/5/92

Modified:
 * 8/8/94 JC
 *	- ANSIfied
 *	- im_open_local added
 * 16/8/94 JC
 *	- im_malloc() added
 * 22/11/94 JC & KM
 *	- tiff added
 * 1/2/95 JC
 *	- tiff removed again!
 *	- now just im_istiff(), im_tiff2vips() and im_vips2tiff()
 * 	- applications have responsibility for making the translation
 * 26/3/96 JC
 *	- im_open_local() now closes on close, not evalend
 * 14/11/96 JC
 *	- tiff and jpeg added again
 * 	- open for read used im_istiff() and im_isjpeg() to switch
 *	- open for write looks at suffix
 * 23/4/97 JC
 *	- im_strdup() now allows NULL IMAGE parameter
 *	- subsample parameter added to im_tiff2vips()
 * 29/10/98 JC
 *	- byte-swap stuff added
 * 16/6/99 JC
 *	- 8x byte swap added for double/double complex
 *	- better error message if file does not exist
 *	- ignore case when testing suffix for save
 *	- fix im_mmapinrw() return test to stop coredump in edvips if
 *	  unwritable
 * 2/11/99 JC
 *	- malloc/strdup stuff moved to memory.c
 * 5/8/00 JC
 *	- fixes for im_vips2tiff() changes
 * 13/10/00 JC
 *	- ooops, missing 'else'
 * 22/11/00 JC
 * 	- ppm read/write added
 * 12/2/01 JC
 * 	- im__image_sanity() added
 * 16/5/01 JC
 *	- now allows RW for non-native byte order, provided it's an 8-bit
 *	  image
 * 11/7/01 JC
 *	- im_tiff2vips() no longer has subsample option
 * 25/3/02 JC
 *	- better im_open() error message
 * 12/12/02 JC
 *	- sanity no longer returns errors, just warns
 * 28/12/02 HB
 *     - Added PNG support
 * 6/5/03 JC
 *	- added im_open_header() (from header.c)
 * 22/5/03 JC
 *	- im__skip_dir() knows about ':' in filenames for vips2tiff etc.
 * 27/11/03 JC
 *	- im_image_sanity() now insists x/y/bands all >0
 * 6/1/04 JC
 *	- moved util funcs out to util.c
 * 18/2/04 JC
 *	- added an im_init_world() to help old progs
 * 16/4/04
 *	- oop, im_open() didn't know about input options in filenames
 * 2/11/04 
 *	- im_open( "r" ) is now lazier ... for non-VIPS formats, we delay 
 *	  read until the first ->generate()
 *	- so im_open_header() is now a no-op
 * 22/6/05
 *	- if TIFF open fails, try going through libmagick
 * 4/8/05
 *	- added analyze read
 * 30/9/05
 * 	- oops, lazy read error recovery didn't clear a pointer
 * 1/5/06
 *	- added OpenEXR read
 * 9/6/06
 * 	- added CSV read/write
 * 20/9/06
 * 	- test for NULL filename/mode, common if you forget to check argc
 * 	  (thanks bruno)
 * 7/11/07
 * 	- use preclose, not evalend, for delayed save
 * 	- add simple cmd-line progress feedback
 */

/*

    This file is part of VIPS.
    
    VIPS is free software; you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */

/*

    These files are distributed with VIPS - http://www.vips.ecs.soton.ac.uk

 */

/*
#define DEBUG_IO
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif /*HAVE_CONFIG_H*/
#include <vips/intl.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#include <vips/vips.h>
#include <vips/debug.h>

#ifdef WITH_DMALLOC
#include <dmalloc.h>
#endif /*WITH_DMALLOC*/

/* Suffix sets.
 */
static const char *im_suffix_vips[] = {
	"v", "",
	NULL
};
static const char *im_suffix_tiff[] = {
	"tif", "tiff",
	NULL
};
static const char *im_suffix_jpeg[] = {
	"jpeg", "jpg", "jfif", "jpe",
	NULL
};
static const char *im_suffix_ppm[] = {
	"ppm", "pbm", "pgm", 
	NULL
};
static const char *im_suffix_png[] = {
	"png", 
	NULL
};
static const char *im_suffix_csv[] = {
	"csv", 
	NULL
};

/* Progress feedback. Only really useful for testing, tbh.
 */
int im__progress = 0;

void
im_progress_set( int progress )
{
	im__progress = progress;
}

/* Open a VIPS image and byte-swap the image data if necessary.
 */
static IMAGE *
read_vips( const char *filename )
{
	IMAGE *im, *im2;

	if( !(im = im_init( filename )) )
		return( NULL );
	if( im_openin( im ) ) {
		im_close( im );
		return( NULL );
	}

	/* Already in native format?
	 */
	if( im_isMSBfirst( im ) == im_amiMSBfirst() ) 
		return( im );

	/* Not native ... but maybe does not need swapping? 
	 */
	if( im->Coding == IM_CODING_LABQ )
		return( im );
	if( im->Coding != IM_CODING_NONE ) {
		im_close( im );
		im_error( "im_open", _( "unknown coding type" ) );
		return( NULL );
	}
	if( im->BandFmt == IM_BANDFMT_CHAR || im->BandFmt == IM_BANDFMT_UCHAR )
		return( im );
	
	/* Needs swapping :( make a little pipeline up to do this for us.
	 */
	if( !(im2 = im_open( filename, "p" )) )
		return( NULL );
        if( im_add_close_callback( im2, 
		(im_callback_fn) im_close, im, NULL ) ) {
                im_close( im );
                im_close( im2 );
                return( NULL );
        }
	if( im_copy_swap( im, im2 ) ) {
		im_close( im2 );
		return( NULL );
	}

	return( im2 );
}

/* Delayed save: if we write to TIFF or to JPEG format, actually do the write
 * to a "p" and on preclose do im_vips2tiff() or whatever. Track save
 * parameters here.
 */
typedef struct {
	int (*save_fn)();	/* Save function */
	IMAGE *im;		/* Image to save */
	char *filename;		/* Save args */
} SaveBlock;

/* From preclose callback: invoke a delayed save.
 */
static int
invoke_sb( SaveBlock *sb )
{
	if( sb->save_fn( sb->im, sb->filename ) )
		return( -1 );

	return( 0 );
}

/* Attach a SaveBlock to an image.
 */
static int
attach_sb( IMAGE *out, int (*save_fn)(), const char *filename )
{
	SaveBlock *sb = IM_NEW( out, SaveBlock );

	if( !sb )
		return( -1 );
	sb->im = out;
	sb->save_fn = save_fn;
	sb->filename = im_strdup( out, filename );

	if( im_add_preclose_callback( out, 
		(im_callback_fn) invoke_sb, (void *) sb, NULL ) )
		return( -1 );

	return( 0 );
}

/* Lazy open: init a descriptor from a filename (just read the header) but
 * delay actually decoding pixels until the first call to a start function.
 */

typedef int (*OpenLazyFn)( const char *filename, IMAGE *im );

/* What we track during a delayed open.
 */
typedef struct _OpenLazy {
	char *filename;

	OpenLazyFn read_pixels;	/* Read in pixels with this */
	IMAGE *lazy_im;		/* Image we read to .. copy from this */
} OpenLazy;

/* Our start function ... do the lazy open, if necessary, and return a region
 * on the new image.
 */
static void *
open_lazy_start( IMAGE *out, void *a, void *dummy )
{
	OpenLazy *lazy = (OpenLazy *) a;

	if( !lazy->lazy_im ) {
		if( !(lazy->lazy_im = im_open_local( out, "read", "p" )) || 
			lazy->read_pixels( lazy->filename, lazy->lazy_im ) ) {
			IM_FREEF( im_close, lazy->lazy_im );
			return( NULL );
		}
	}

	return( im_region_create( lazy->lazy_im ) );
}

/* Just copy.
 */
static int
open_lazy_generate( REGION *or, void *seq, void *a, void *b )
{
	REGION *ir = (REGION *) seq;

        Rect *r = &or->valid;

        /* Ask for input we need.
         */
        if( im_prepare( ir, r ) )
                return( -1 );

        /* Attach output region to that.
         */
        if( im_region_region( or, ir, r, r->left, r->top ) )
                return( -1 );

        return( 0 );
}

/* Lazy open ... init the header with the first OpenLazyFn, delay actually
 * decoding pixels with the second OpenLazyFn until the first generate().
 */
static int
open_lazy( OpenLazyFn read_header, OpenLazyFn read_pixels, 
	const char *filename, IMAGE *out )
{
	OpenLazy *lazy = IM_NEW( out, OpenLazy );

	if( !lazy ||
		!(lazy->filename = im_strdup( out, filename )) )
		return( -1 );
	lazy->read_pixels = read_pixels;
	lazy->lazy_im = NULL;

	if( read_header( filename, out ) )
		return( -1 );

	if( im_generate( out, 
		open_lazy_start, open_lazy_generate, im_stop_one, 
		lazy, NULL ) )
		return( -1 );

	return( 0 );
}

static IMAGE *
open_sub( OpenLazyFn read_header, OpenLazyFn read_pixels, const char *filename )
{
	IMAGE *im;

	if( !(im = im_open( filename, "p" )) || 
		open_lazy( read_header, read_pixels, filename, im ) ) {
		im_close( im );
		return( NULL );
	}

	return( im );
}

/* Progress feedback. 
 */

/* What we track during an eval.
 */
typedef struct {
	IMAGE *im;

	int last_percent;	/* The last %complete we displayed */
} Progress;

int
evalstart_cb( Progress *progress )
{
	progress->last_percent = 0;

	return( 0 );
}

int
eval_cb( Progress *progress )
{
	IMAGE *im = progress->im;

	if( im->time->percent != progress->last_percent ) {
		/* \r at end returns to the start of this line.
		 */
		printf( _( "%s: %d%% complete\r" ), 
			im->filename, im->time->percent );
		fflush( stdout );

		progress->last_percent = im->time->percent;
	}

	return( 0 );
}

int
evalend_cb( Progress *progress )
{
	IMAGE *im = progress->im;

	/* Spaces at end help to erase the %complete message we overwrite.
	 */
	printf( _( "%s: done in %ds          \n" ), 
		im->filename, im->time->run );

	return( 0 );
}

IMAGE *
im_open( const char *filename, const char *mode )
{
	IMAGE *im;

	/* Pass in a nonsense name for argv0 ... this init world is only here
	 * for old programs which are missing an im_init_world() call. We must
	 * have threads set up before we can process.
	 */
	if( im_init_world( "vips" ) )
		im_error_clear();

	if( !filename || !mode ) {
		im_error( "im_open", _( "NULL filename or mode" ) );
		return( NULL );
	}

	switch( mode[0] ) {
        case 'r':
{
		char name[FILENAME_MAX];
		char options[FILENAME_MAX];

		/* Break any options off the name ... eg. "fred.tif:jpeg,tile" 
		 * etc.
		 */
		im_filename_split( filename, name, options );

		/* Check for other formats.

			FIXME ... should have a table to avoid all this
			repetition

		 */
		if( !im_existsf( "%s", name ) ) {
			im_error( "im_open", 
				_( "\"%s\" is not readable" ), name );
			return( NULL );
		}
		else if( im_istiff( name ) ) {
			/* If TIFF open fails, try going through libmagick.
			 */
			if( !(im = open_sub( 
				im_tiff2vips_header, im_tiff2vips, 
					filename )) &&
				!(im = open_sub( 
					im_magick2vips_header, im_magick2vips, 
					filename )) )
				return( NULL );
		}
		else if( im_isjpeg( name ) ) {
			if( !(im = open_sub( 
				im_jpeg2vips_header, im_jpeg2vips, filename )) )
				return( NULL );
		}
		else if( im_isexr( name ) ) {
			if( !(im = open_sub( 
				im_exr2vips_header, im_exr2vips, filename )) )
				return( NULL );
		}
		else if( im_isppm( name ) ) {
			if( !(im = open_sub( 
				im_ppm2vips_header, im_ppm2vips, filename )) )
				return( NULL );
		}
		else if( im_ispng( name ) ) {
			if( !(im = open_sub( 
				im_png2vips_header, im_png2vips, filename )) )
				return( NULL );
		}
		else if( im_filename_suffix_match( name, im_suffix_csv ) ) {
			if( !(im = open_sub( 
				im_csv2vips_header, im_csv2vips, filename )) )
				return( NULL );
		}
		else if( im_isvips( name ) ) {
			if( mode[1] == 'w' ) {
				/* Has to be native format for >8 bits.
				 */
				if( !(im = im_init( filename )) )
					return( NULL );
				if( im_openinrw( im ) ) {
					im_close( im );
					return( NULL );
				}
				if( im->Bbits != IM_BBITS_BYTE &&
					im_isMSBfirst( im ) != 
						im_amiMSBfirst() ) {
					im_close( im );
					im_error( "im_open", _( "open for read-"
						"write for native format "
						"images only" ) );
					return( NULL );
				}
			}
			else 
				im = read_vips( filename );
		}
		else if( im_isanalyze( name ) ) {
			if( !(im = open_sub( 
				im_analyze2vips_header, im_analyze2vips, 
					filename )) )
				return( NULL );
		}
		else if( im_ismagick( name ) ) {
			/* Have this last as it can be very slow to detect
			 * failure.
			 */
			if( !(im = open_sub( 
				im_magick2vips_header, im_magick2vips, 
				filename )) )
				return( NULL );
		}
		else {
			im_error( "im_open", _( "\"%s\" is not "
				"a supported format" ), filename );
			return( NULL );
		}
}

        	break;

	case 'w':
		/* Look at the suffix for format write.
		 */
		if( im_filename_suffix_match( filename, im_suffix_vips ) ) 
			im = im_openout( filename );
		else if( im_filename_suffix_match( filename, 
			im_suffix_tiff ) ) {
			/* TIFF write. Save to a partial, and on preclose
			 * im_vips2tiff from that.
			 */
			if( !(im = im_open( "im_open:vips2tiff:1", "p" )) )
				return( NULL );
			if( attach_sb( im, im_vips2tiff, filename ) ) {
				im_close( im );
				return( NULL );
			}
		}
		else if( im_filename_suffix_match( filename, 
			im_suffix_jpeg ) ) {
			/* JPEG write. 
			 */
			if( !(im = im_open( "im_open:vips2jpeg:1", "p" )) )
				return( NULL );
			if( attach_sb( im, im_vips2jpeg, filename ) ) {
				im_close( im );
				return( NULL );
			}
		}
		else if( im_filename_suffix_match( filename, im_suffix_ppm ) ) {
			/* ppm write. 
			 */
			if( !(im = im_open( "im_open:vips2ppm:1", "p" )) )
				return( NULL );
			if( attach_sb( im, im_vips2ppm, filename ) ) {
				im_close( im );
				return( NULL );
			}
		}
		else if( im_filename_suffix_match( filename, im_suffix_png ) ) {
			/* png write. 
			 */
			if( !(im = im_open( "im_open:vips2png:1", "p" )) )
				return( NULL );
			if( attach_sb( im, im_vips2png, filename ) ) {
				im_close( im );
				return( NULL );
			}
		}
		else if( im_filename_suffix_match( filename, im_suffix_csv ) ) {
			/* csv write. 
			 */
			if( !(im = im_open( "im_open:vips2csv:1", "p" )) )
				return( NULL );
			if( attach_sb( im, im_vips2csv, filename ) ) {
				im_close( im );
				return( NULL );
			}
		}
		else {
			im_error( "im_open", _( "unknown suffix for save" ) );
			return( NULL );
		}
        	break;

        case 't':
                im = im_setbuf( filename );
                break;

        case 'p':
                im = im_partial( filename );
                break;

	default:
		/* Getting here means bad mode
		 */
		im_error( "im_open", _( "bad mode \"%s\"" ), mode );
		return( NULL );
        }

	/* Attach progress feedback, if required.
	 */
	if( im__progress ) {
		Progress *progress = IM_NEW( im, Progress );

		progress->im = im;
		im_add_evalstart_callback( im, 
			(im_callback_fn) evalstart_cb, progress, NULL );
		im_add_eval_callback( im, 
			(im_callback_fn) eval_cb, progress, NULL );
		im_add_evalend_callback( im, 
			(im_callback_fn) evalend_cb, progress, NULL );
	}

#ifdef DEBUG_IO
	printf( "im_open: success for %s (%p)\n", im->filename, im );
#endif /*DEBUG_IO*/

	return( im );
}

/* Test an image for sanity. We could add many more tests here.
 */
static const char *
image_sanity( IMAGE *im )
{
	if( !im ) 
		return( "NULL descriptor" );
	if( !im->filename ) 
		return( "NULL filename" );
	if( !g_slist_find( im__open_images, im ) )
		return( "not on open image list" );

	if( im->Xsize <= 0 || im->Ysize <= 0 || im->Bands <= 0 ) 
		return( "bad dimensions" );
	if( im->BandFmt < -1 || im->BandFmt > IM_BANDFMT_DPCOMPLEX ||
		(im->Coding != -1 &&
			im->Coding != IM_CODING_NONE && 
			im->Coding != IM_CODING_LABQ) ||
		im->Type < -1 || im->Type > IM_TYPE_GREY16 ||
		im->dtype > IM_PARTIAL || im->dhint > IM_ANY ) 
		return( "bad enum value" );
	if( im->Xres < 0 || im->Xres < 0 ) 
		return( "bad resolution" );

	return( NULL );
}

int 
im_image_sanity( IMAGE *im )
{
	const char *msg;

	if( (msg = image_sanity( im )) ) {
		im_warn( "im_image_sanity", "\"%s\" (%p) %s",
			im ? (im->filename ? im->filename : "") : "", 
			im, msg );
		im_printdesc( im );
	}

	return( 0 );
}

/* Just here for compatibility.
 */
IMAGE *
im_open_header( const char *file )
{
	return( im_open( file, "r" ) );
}
